### PR TITLE
Fix existing branch bug and extend Changelog compatibility

### PR DESCRIPTION
**Pull request by mxkyb**
## Existing Branches

When branches were already present, previously the action would fail. Now these branches are updated.

## Changelog

Some Changelogs use a format like this:
```
## [1.2.3] - Unreleased

- some
- changes
```
while others use:
```
## Unreleased

- some
- changes
```
This pull request introduces parsing compatibility for both formats and will always output the first.

### CODE REVIEW

File: dist/index.js [hunk 1]
```diff
@@ -26077,25 +26077,37 @@ const getPayload = () => {
...
+        owner,
+        repo,
         branch,
-        ...rest
+        sha,
     } = data;
-    
+
+    try {
+        await octokit.rest.git.getRef({
```
Comment (gotbahn) [inline dist/index.js:26087]: why do you need get something you're not using?

File: src/github-api.js [hunk 2]
```diff
@@ -24,12 +24,30 @@ export const getPayload = () => {
...
+            repo,
+            ref: `heads/${branch}`,
+        });
+    } catch {}
+
     await octokit.rest.git.createRef({
-        ...rest,
+        owner,
+        repo,
+        sha,
```
Comment (nitzanashi) [inline src/github-api.js:50]: can't we use `updateRef` with `force` option?
https://docs.github.com/en/rest/git/refs#update-a-reference

File: src/index.js [hunk 3]
```diff
@@ -58,7 +59,7 @@ const editChangelog = async ({
 
     const date = format(new Date(), "dd.MM.yyyy")
 
-    const changelogDateCut = rawChangelog.replace('Unreleased', date);
+    const changelogDateCut = rawChangelog.replace(/##(\s\[.+\]\s\-)?(\sUnreleased)/, `## [${releaseVersion}] - ${date}`);
```
Comment (nitzanashi) [inline src/index.js:62]: how is it different to 
https://github.com/zattoo/changelog/blob/master/src/validate.js#L19

File: dist/index.js (hunk 1 above)
Comment (nitzanashi) [reply dist/index.js:26087]: True that, why can't we just try to delete ref?

File: src/github-api.js (hunk 2 above)
Comment (mxkyb) [reply src/github-api.js:50]: `deleteRef` will also close pull requests associated with branches, which i like. other than that, we could also use `updateRef`

File: src/index.js (hunk 3 above)
Comment (mxkyb) [reply src/index.js:62]: its much simpler

it doesn't support:
```
## [Unreleased]
```
and also doesn't support anything that doesn't have `Unreleased` in it

it does support:
```
## Unreleased
## [3.2026.2] - Unreleased
```

File: src/github-api.js (hunk 2 above)
Comment (mxkyb) [reply src/github-api.js:50]: When cut is run twice, the second run will update the branches and then fail when using `updateRef`.
With `deleteRef` it will update the branches and recreate the pull requests.
probably so much edge case that it doesn't really matter

File: dist/index.js [hunk 4]
```diff
@@ -26077,25 +26077,33 @@ const getPayload = () => {
...
+            sha,
+        });
+    } catch {
+        await octokit.rest.git.createRef({
+            owner,
+            repo,
+            sha,
+            ref: `refs/heads/${branch}`,
+        });
+    }
```
Comment (nitzanashi) [inline dist/index.js:26101]: I don't really understand this handeling of error,
if we fail to `updateRef` it might be from many reasons, but what about failing of `createRef` We need to pass a meaningful error to the users, also that we can debug it better, In general i'd recommand to add a bit more meaningful logging to the action

File: src/github-api.js [hunk 5]
```diff
@@ -24,14 +24,28 @@ export const getPayload = () => {
...
+            sha,
+        });
+    } catch {
+        await octokit.rest.git.createRef({
+            owner,
+            repo,
+            sha,
+            ref: `refs/heads/${branch}`,
+        });
+    }
```
Comment (nitzanashi) [inline src/github-api.js:48]: I don't really understand this handeling of error,
if we fail to `updateRef` it might be from many reasons, but what about failing of `createRef` We need to pass a meaningful error to the users, also that we can debug it better, In general i'd recommand to add a bit more meaningful logging to the action

Comment (mxkyb) [reply src/github-api.js]: like this? https://github.com/zattoo/web-shadow/actions/runs/2571203870

Comment (mxkyb) [reply src/github-api.js]: also, i separated `createRef` from `updateRef` without introducing other checks like for 404, because it seems like a very edge case

**Comment by nitzanashi** [comment]
@mxkyb in general looks good, but does it work? any test references?

**Comment by nitzanashi** [comment]
> this is a success when closing the pull request: https://github.com/zattoo/web-shadow/actions/runs/2571213000

Why success say "fail" ?

image

**Comment by mxkyb** [comment]
@nitzanashi we can parse the error response, will do. 

It says fail because we don’t know why we are in the catch block. Calling update ref after is more because it is the most likely fix to the problem, but not because we actually know it’s the problem. We can explicitly check the error response for the code for branch already exists and make the error output more clear

**Comment by mxkyb** [comment]
example PRs:
- https://github.com/zattoo/web-shadow/pull/404
- https://github.com/zattoo/web-shadow/pull/405

**Comment by mxkyb** [comment]
@nitzanashi requesting review by comment (control not working)

File: src/index.js [hunk 6]
```diff
@@ -224,6 +231,7 @@ const createReleaseCandidatePullRequest = async ({
     const token = core.getInput('token', {required: true});
     const rcLabels = core.getMultilineInput('labels', {required: false});
     const project = core.getInput('project', {required: true});
+    const releaseVersionInput = core.getInput('release_version', {required: false});
     const nextVersion = core.getInput('next_version', {required: true});
```
Comment (nitzanashi) [inline src/index.js:235]: next version might be unknonwn in projects like account seo embeds (anything that doesn't follow train)

Comment (mxkyb) [reply src/index.js:235]: Hm, I had hoped we can switch them to our way of handling releases. I will make it optional then and support both cutting patterns

Comment (nitzanashi) [reply src/index.js:235]: Not at all we're gonna go the other way around in the future to pseudo versions on the main branch (e.g. 0.0.0) and defintion of the release will happen on cut time

Comment (mxkyb) [reply src/index.js:235]: this will ease the next and current version confusion. But for now we need to support both patterns?

Comment (mxkyb) [reply src/index.js]: @nitzanashi added changes

File: CHANGELOG.md [hunk 7]
```diff
@@ -3,6 +3,12 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](http://keepachangelog.com/)
 
+## [4.0.1] - 23.06.2022
```
Comment (nitzanashi) [inline CHANGELOG.md:6]:
```suggestion
## [4.1.0] - 23.06.2022
```

It is a bit bigger than fixing a bug hahaha

Comment (mxkyb) [reply CHANGELOG.md]: changed it

**Comment by mxkyb** [comment]
example prs here: https://github.com/zattoo/web-shadow/pulls

File: src/index.js [hunk 8]
```diff
@@ -122,18 +138,20 @@ const createVersionRaisePullRequest = async ({
 
     const updatedFiles = {
         packageJson: editPackageJson({
-            nextVersion,
+            version: nextVersion || releaseVersion,
```
Comment (nitzanashi) [inline src/index.js:141]: cache it
`const version = nextVersion || releaseVersion;`

Comment (mxkyb) [reply src/index.js]: changed

**Review by nitzanashi — APPROVED** [review]
